core/src/main/java/com/datastrato/gravitino/lock/TreeLockNode.java

```diff
@@ -0,0 +1,136 @@
+/*
+ * Copyright 2024 Datastrato Pvt Ltd.
+ * This software is licensed under the Apache License version 2.
+ */
+
+package com.datastrato.gravitino.lock;
+
+import com.datastrato.gravitino.NameIdentifier;
+import com.google.common.collect.Lists;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+import javax.annotation.concurrent.ThreadSafe;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+@ThreadSafe
+public class TreeLockNode {
+  public static final Logger LOG = LoggerFactory.getLogger(TreeLockNode.class);
+  private final NameIdentifier ident;
+  private final ReentrantReadWriteLock readWriteLock;
+  private final Map<NameIdentifier, TreeLockNode> childMap;
+  private final LockManager lockManager;
+
+  private final AtomicLong referenceCount = new AtomicLong();
+
+  public TreeLockNode(NameIdentifier identifier, LockManager manager) {
+    this.ident = identifier;
+    this.readWriteLock = new ReentrantReadWriteLock();
+    this.childMap = new ConcurrentHashMap<>();
+    this.lockManager = manager;
+  }
+
+  public NameIdentifier getIdent() {
+    return ident;
+  }
+
+  // Why is this method synchronized, please see the comment in the method
+  // LockManager#evictStaleNodes
+  public synchronized void addReference() {
+    referenceCount.getAndIncrement();
+  }
+
+  public void decReference() {
+    referenceCount.getAndDecrement();
+  }
+
+  public long getReferenceCount() {
+    return referenceCount.get();
+  }
+
+  /**
+   * Try to lock the node with the given lock type.
+   *
+   * @param lockType The lock type to lock the node.
+   * @return True if the node is locked successfully, false otherwise.
+   */
+  public boolean tryLock(LockType lockType) {
+    boolean success =
+        lockType == LockType.READ
+            ? readWriteLock.readLock().tryLock()
+            : readWriteLock.writeLock().tryLock();
+    if (success) {
+      referenceCount.getAndIncrement();
+
+      LOG.info(
+          "Node {} has been lock with '{}' lock, reference count = {}",
+          ident,
+          lockType,
+          referenceCount.get());
+    }
+
+    return success;
```

Comment:
You should add more tracing log here and below in lock/unlock.

Reply:
Added